evap/staff/importers/user.py

```diff
@@ -0,0 +1,384 @@
+
+        self.importer_log.add_warning(msg, category=ImporterLogEntry.Category.DUPL)
+
+    @staticmethod
+    def _create_user_string(user: Union[UserProfile, UserData]):
+        if isinstance(user, UserProfile):
+            return format_html(
+                "{} {} {}, {} [{}]",
+                user.title or "",
+                user.first_name,
+                user.last_name,
+                user.email or "",
+                user_edit_link(user.pk),
+            )
+
+        return format_html("{} {} {}, {}", user.title or "", user.first_name, user.last_name, user.email or "")
+
+
+class UserDataValidationChecker(Checker):
+    """Run django validation against UserData instances"""
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        # Only give one error per unique user_data.
+        self.already_checked: Set[UserData] = set()
+
+    def check_userdata(self, user_data: UserData, _location: ExcelFileLocation):
+        if user_data.email == "":
+            # Should trigger another checker. We cannot meaningfully give an error message.
+            return
+
+        if user_data in self.already_checked:
+            return
+        self.already_checked.add(user_data)
+
+        try:
+            user_data.validate()
+        except ValidationError as e:
+            self.importer_log.add_error(
+                _("User {user_email}: Error when validating: {error}").format(user_email=user_data.email, error=e),
+                category=ImporterLogEntry.Category.USER,
+            )
+
+
+class DuplicateUserDataChecker(Checker):
+    """Check for duplicate users"""
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.first_location_by_user_data: Dict[UserData, ExcelFileLocation] = {}
+
+        self.tracker: FirstLocationAndCountTracker = FirstLocationAndCountTracker()
+
+    def check_userdata(self, user_data: UserData, location: ExcelFileLocation):
+        if user_data not in self.first_location_by_user_data:
+            self.first_location_by_user_data[user_data] = location
+            return
+
+        first_location = self.first_location_by_user_data[user_data]
+        self.tracker.add_location_for_key(location=location, key=first_location)
+
+    def finalize(self):
+        for first_location, location_string in self.tracker.aggregated_keys_and_location_strings():
+            self.importer_log.add_warning(
+                _("{location}: The duplicated row was ignored. It was first found at {first_location}.").format(
+                    location=location_string,
+                    first_location=first_location,
+                ),
+                category=ImporterLogEntry.Category.IGNORED,
+            )
+
+
+class UserDataAdapter(RowCheckerMixin):
+    """Adapter to use Checkers for UserData with UserParsedRow"""
+
+    def __init__(self, user_data_checker):
```

Comment:
```suggestion
    def __init__(self, user_data_checker: Checker):
```

Reply:
Doesn't work with the current code base: `Checker` only defined `__init__` and `finalize`, so we can have checkers for `UserData`, `CourseData`, different kinds of rows, ...

We could add a protocol for `UserDataChecker`, but then we'd probably have to add protocols for all other cases, too?

Reply:
Hm, both protocols and generics _could_ make this work, but we need to draw the line somewhere and I am fine with doing that here - otherwise I have to open PRs at mypy